encoding/string_encode.go

```diff
@@ -1,30 +1,182 @@
+		},
+		data: make([]string, defaultInitalDataSliceSize),
+	}
+}
 
 // Encode encodes a collection of strings and writes the encoded bytes to the writer.
-func (enc *StringEnc) Encode(writer io.Writer, values ForwardStringIterator) error {
-	return errors.New("not implemented")
+func (enc *StringEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableStringIterator,
+) error {
+	// TODO(bodu): Do some perf benchmarking to see whether we want to allocate a new map
+	// or clear an existing one.
+	dictionary := make(map[string]int64)
+	var (
+		idx       int64
+		curr      string
+		maxLength int
+	)
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		if len(dictionary) < dictEncodingMaxCardinality {
+			// Only add to dictionary if
+			if _, ok := dictionary[curr]; !ok {
+				dictionary[curr] = idx
+				// Ensure data size or grow slice. Target size is idx+1.
+				enc.data = enc.ensureStringSliceSize(enc.data, int(idx)+1)
+				enc.data[idx] = curr
+				idx++
+			}
+		}
+		maxLength = int(math.Max(float64(len(curr)), float64(maxLength)))
+	}
+
+	if err := valuesIt.Err(); err != nil {
+		return err
+	}
+
+	// Rewind values.
+	valuesIt.Rewind()
+
+	// TODO(bodu): This should take into account the total # of items
+	// at some point. The total count should exceed the # of uniques by a certain
+	// margin to justify table compression.
+	if len(dictionary) >= dictEncodingMaxCardinality {
+		enc.metaProto.Encoding = encodingpb.EncodingType_RAW_SIZE
+	} else {
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+	}
+
+	enc.buf = bytes.EnsureBufferSize(enc.buf, enc.metaProto.Size(), bytes.DontCopyData)
+	if err := ProtoEncode(&enc.metaProto, enc.buf, writer); err != nil {
+		return err
+	}
+
+	var compressWriter io.Writer
+	// Compress the bytes.
+	switch enc.metaProto.Compression {
+	case encodingpb.CompressionType_ZSTD:
+		// TODO(bodu): Figure out a cleaner way to do this.
+		w := gozstd.NewWriter(writer)
+		defer w.Close()
+		compressWriter = w
+	default:
+		return fmt.Errorf("invalid compression type: %v", enc.metaProto.Compression)
+	}
+
+	switch enc.metaProto.Encoding {
+	case encodingpb.EncodingType_RAW_SIZE:
+		if err := enc.encodeLength(compressWriter, valuesIt); err != nil {
+			return err
+		}
+	case encodingpb.EncodingType_DICTIONARY:
+		if err := enc.encodeDictionary(compressWriter, valuesIt, dictionary, enc.data[:idx]); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("invalid encoding type: %v", enc.metaProto.Encoding)
+	}
+
+	return nil
 }
 
-// Reset resets the encoder.
-func (enc *StringEnc) Reset() { panic("not implemented") }
+// Dictionary encoding strategy is to write
+// and then the bytes for the string.
+func (enc *StringEnc) encodeDictionary(
+	writer io.Writer,
+	valuesIt RewindableStringIterator,
+	dictionary map[string]int64,
+	data []string,
+) error {
+	// Write out the dictionary data first (to be used for decoding).
+	enc.dictionaryProto.Data = data
```

Comment:
Perhaps call `enc.dictionaryProto.Data = enc.dictionaryProto.Data[:0]` before setting data for reusing across calls?

Reply:
What does that do besides restarting the slice header to the 0 idx?

I'm replacing the slice header with an entirely new header.

Reply:
Yeah np misread it, ignore it.